internal/v4/search.go

```diff
@@ -130,6 +134,20 @@ func parseSearchParams(req *http.Request) (charmstore.SearchParams, error) {
 			return charmstore.SearchParams{}, badRequestf(nil, "invalid parameter: %s", k)
 		}
 	}
-
 	return sp, nil
 }
+
+func (h *Handler) groupsFromRequest(sp *charmstore.SearchParams, req *http.Request) {
+	ms := httpbakery.RequestMacaroons(req)
+	if len(ms) == 0 {
+		return
+	}
+	dec, err := httpbakery.CheckRequest(h.store.Bakery, req, nil, checkers.New())
```

Comment:
I'm a bit uncomfortable about this being separate from the main auth code and without constants for the username and groups attributes, if those change, this becomes a potential security information leak.

I suggest creating a Handler.checkRequest method that returns a struct holding the declared attributes (only the ones that we added to the need-declared caveat) and an error, and using it here.